### PR TITLE
Span class.

### PR DESCRIPTION
**Pull request by trivialfis**
Implement span class from ISO++20 for safer pointer indexing.

TODOs:
- [x] Fix convert and container constructor.
- [x] Fix as_bytes.
- [x] Structure tests for CUDA tests.
- [x] Fix errors in test farm.
- [x] Make an example use in xgboost.
- [x] Make it work with MSVC.

### CODE REVIEW

File: src/common/span.h [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+  XGBOOST_DEVICE const_reverse_iterator crbegin() {
+    return const_reverse_iterator(cend());
+  }
+  XGBOOST_DEVICE const_reverse_iterator crend() {
+    return const_reverse_iterator{cbegin()};
+  }
+
+  XGBOOST_DEVICE reference operator[](index_type idx) const {
+    SPAN_CHECK(idx >= 0 && idx < size());
+    SPAN_CHECK(idx >= 0 && idx < size());
```
Comment (RAMitchell) [inline src/common/span.h]: Duplicate check here.

File: tests/cpp/common/test_span.cc [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+
+TEST(Span, TestStatus) {
+  int status = 1;
+  TestTestStatus {&status}();
+  ASSERT_EQ(status, -1);
+}
+
+TEST(Span, Constructors) {
+  // Dynamic extent
+  {
```
Comment (RAMitchell) [inline tests/cpp/common/test_span.cc]: These two tests are identical.

File: tests/cpp/common/test_span.cc [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+  }
+}
+
+TEST(Span, FromNullPtr) {
+  // dynamic extent
+  {
+    Span<float> s {nullptr, static_cast<Span<float>::index_type>(0)};
+    ASSERT_EQ(s.size(), 0);
+    ASSERT_EQ(s.data(), nullptr);
+
```
Comment (RAMitchell) [inline tests/cpp/common/test_span.cc]: More duplicated tests.

**Comment by trivialfis** [comment]
@RAMitchell  I will re-visit all test cases, see what's been missing.
For the as_bytes, it did came across my mind. Maybe one needs to serialize a segment of memory, like a trained model? I don't know, so obliging to the standard is just me making a safe bet.

**Comment by trivialfis** [comment]
Not ready yet, seems tests is still not thorough enough.

**Comment by RAMitchell** [comment]
@hcho3 as noted by @trivialfis there is actually an array_view class in dmlc-core. After a brief look it seems it is used by the nvvm/tv projects. We could replace array_view with our span which is more powerful,  better tested and compliant with the upcoming standard. What do you think?

**Comment by hcho3** [comment]
@RAMitchell You should file a pull request at dmlc-core and make a case for upgrading array_view with the span class. There seems to be a good amount of code using array_view already and it would take some work to replace all occurrences.

For the sake of lesser friction, you could simply add the span class to dmlc-core and have DMLC projects gracefully adopt it.

**Comment by trivialfis** [comment]
@RAMitchell Ready for another review. The Travis failed fetching packages in Python test, it should be fine on a good day.

The last commit is just a toy show case for using Span in kernel code. It could be reverted for a more formal integration.

**Comment by RAMitchell** [comment]
Looks good. One more thing to consider, span will incur a performance overhead when using the operator[], these changes will have worsened performance in places like this: https://github.com/dmlc/xgboost/blob/a96039141a8dd2d053ae8f415c8fc8e52223216e/src/data/simple_dmatrix.cc#L61

This can be fixed by changing this to a range based for loop, I believe this does not need to do any index checking.

**Comment by RAMitchell** [comment]
DVec can be dealt with later. My idea is that we can gradually enforce the usage of span across the code base using the clang-tidy linter. The most important algorithm for us is gpu_hist as it is used a lot in production, so maybe we can start there when you are ready.

I think this should be good to merge when the CI tests pass.

**Comment by trivialfis** [comment]
@RAMitchell So it passed. \0/

**Pull request merged**